submissions/angels_unit_tests/angels_unit_tests.md

```diff
@@ -0,0 +1,620 @@
+
+## It started small
+
+With the great help of the community we learned how the Angel's libary worked. For those that do not know what a library is, it is a set of helper functions that abstract the factorio code base such that it becomes easier to program. It is mainly a list with quality of life additions for the developer, such that he can focus on developing features, rather than writing boiler plate code to get something to work. You can compare Angel's library to some stand-alone mods like the [Factorio Library](https://mods.factorio.com/mod/flib), [Bob's Functions Library mod](https://mods.factorio.com/mod/boblibrary) or even my own attempt [Lovely_santa's library of knowledge](https://mods.factorio.com/mod/LSlib). There is no 'right' or 'wrong' way to implement a library, as long as it suits its purpose. By understanding Angel's library, which actually lives inside Angel's refining and not as a stand-alone mod, we figured more efficient ways of dealing with certain bugs arrising in the mod.
+
+I am confident now saying that we do not make coding mistakes anymore, code slipping through the cracks, which would result in a hard crash of the mods. Nowadays the issues are likely edge cases, unhandled cases or a strange modpack configuration. Most frequently we receive bugs from people who play with a few Angel's mods and a few Bob mods. The reason for this is simple: there are many configurations with many settings, which result in many oversights. [A small example](https://www.reddit.com/r/factorio/comments/pfvqn0/playing_an_angel_bobs_game_and_noticed_about_9/) is shown below; you fix a bug with some technology being unresearchable, in some exotic configuration:
+
+![Nickel processing is unresearchable](https://raw.githubusercontent.com/LovelySanta/Alt-F4/angels-unit-tests/submissions/angels_unit_tests/nickel_processing_bug_1.png)
+
+After a while you notice another configuration:
+
+![Nickel processing is unresearchable AGAIN!](https://raw.githubusercontent.com/LovelySanta/Alt-F4/angels-unit-tests/submissions/angels_unit_tests/nickel_processing_bug_2.png)
+
+As a solution, we began creating a list of configurations we had to test. [*pezzawinkle*](https://mods.factorio.com/user/pezzawinkle) recently made [a not-so-small list](https://github.com/Arch666Angel/mods/blob/master/Config_Testing_Regime.txt) with all configurations he tests every time he prepairs a release.
```

Comment:
This is a stretch, but is `pezzawinkle` maybe this [guy on YouTube](https://www.youtube.com/c/PezzzasWork/featured)? If so it would be awesome if you could link a video to give some background info info about him. The C++ projects this guy creates are quite amazing.

Reply:
I've asked the question at @Pezzawinkle, no answer yet. Seems like the first time I've seen that youTube page...

Reply:
I can confirm that that page is NOT mine (sadly). If you want a quick background for me. I have a PhD in Chemical Engineering from the University of Sydney, specializing in energy and availability analysis methods. I have been playing factorio and AngelBobs since 0.14. I do have a youtube channel which I experimented with last year during a lockdown, but am unable to continue uploading vids for that, so not really worth linking. I have been working with LovelySanta since Angel gave control to the dev team and have learned a massive amount throughout the process. I am also to blame for pushing out the undercooked versions of component and tech modes as part of angelsindustries back in 0.18.

Reply:
@Pezzawinkle thanks for the input! Haha it was a stretch. But the work you guys are doing on Angels are super impressive in iteself already! @LovelySanta you're welcome to add some of this information into your piece if you want, otherwise I will move I believe this piece is ready :)

Reply:
Nah its fine, we're not that hungry for attention, just wanted to give credit where its due. It is sufficient as is; marking it as resolved.